src/org/labkey/test/components/glassLibrary/heatmap/HeatMap.java

```diff
@@ -0,0 +1,221 @@
+package org.labkey.test.components.glassLibrary.heatmap;
+
+import org.labkey.test.Locator;
+import org.labkey.test.WebDriverWrapper;
+import org.labkey.test.components.Component;
+import org.labkey.test.components.WebDriverComponent;
+import org.openqa.selenium.By;
+import org.openqa.selenium.StaleElementReferenceException;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static org.labkey.test.WebDriverWrapper.WAIT_FOR_JAVASCRIPT;
+import static org.labkey.test.util.TestLogger.log;
+
+/**
+ * Manipulates the HeatMap component from labkey-ui-components (components/src/internal/components/heatmap/HeatMap.tsx)
+ */
+public class HeatMap extends WebDriverComponent<HeatMap.ElementCache>
+{
+    private final WebElement _el;
+    private final WebDriver _driver;
+    private final Locator.XPathLocator _rowsLoc = Locator.tag("tbody").childTag("tr").withoutClass("grid-empty").withoutClass("grid-loading");
+    private final Locator _loadingRowLoc = Locator.css("tbody tr.grid-loading");
+    private final Locator _spinnerLoc = Locator.css("span i.fa-spinner");
+    private final Locator _emptyRowLoc = Locator.css("tbody tr.grid-empty");
+
+    protected HeatMap(WebElement element, WebDriver driver)
+    {
+        _el = element;
+        _driver = driver;
+    }
+
+    @Override
+    public WebElement getComponentElement()
+    {
+        return _el;
+    }
+
+    @Override
+    public WebDriver getDriver()
+    {
+        return _driver;
+    }
+
+    public String getColumnByIndex(int index)
+    {
+        return getColumnNames().get(index);
+    }
+
+    public List<String> getColumnNames()
+    {
+        List<WebElement> headerCells = elementCache().headerCells();
+        return getWrapper().getTexts(headerCells);
+    }
+
+    public List<String> getEntityNames()
+    {
+        List<String> entityNames = new ArrayList<>();
+        for (WebElement row : getRows())
+        {
+            WebElement entityLink = Locator.xpath("//td/div/a").findElement(row);
+            entityNames.add(entityLink.getText());
+        }
+        return entityNames;
+    }
+
+    public Optional<WebElement> getOptionalRow(String linkText)
+    {
+        List<WebElement> rows = getRows();
+        if(rows.size() == 0)  // try again; we expect there to be rows if we're here
+        {
+            new WebDriverWait(getDriver(), 10).until(ExpectedConditions.numberOfElementsToBeMoreThan(By.cssSelector("tbody tr"), 0));
+            rows=getRows();
+        }
+        for (WebElement row : rows)
+        {
+            WebElement firstCell = Locator.tag("td").findElement(row);
+            if (Locator.linkWithText(linkText).existsIn(firstCell))
+            {
+                return Optional.of(row);
+            }
```

Comment:
Squash this into one `Locator`.
```suggestion
            if (Locator.tag("td").append(Locator.linkWithText(linkText)).existsIn(row))
            {
                return Optional.of(row);
            }
```